### ISSUE TITLE
Install/compile fails of letsencrypt-auto on Smartos/Illumos

### ISSUE THREAD

**Issue by fire** [certbot/certbot#1173]
The auto method fails because a script was not written.

Can't seem to find Augeas in the package manager. It may be possible to install letsencrypt client manually.

Seems avaliable since Smartos pkgsrc 2015Q1

Will try to target 2015Q3

**Comment by kuba** [comment]
Augeas is only necessary for `apache` plugin - maybe we could simply not install apache plugin in case of you platform?

**Comment by fire** [comment]
How do you disable the plugin?

**Comment by fire** [comment]
I was able to get the client working after I disabled the Apache plugin and the tests. 

The python on Smartos 14.4.0 is named python2.7. So it had to be renamed from python2.

**Comment by kuba** [comment]
With #1257 you will be able to call `LE_PYTHON=python2.7 letsencrypt-auto`.

If you can provide autodetection for your OS in `letsencrypt-auto` script and set `LE_PYTHON` appropriately, that would be awesome.

**Comment by fire** [comment]
Is that patch in the master branch yet?

**Comment by kuba** [comment]
#1257 has not yet been merged to `master` (it's still green, rather than purple)

**Comment by fire** [comment]
Is there a better way of disabling the apache plugin than just disabling the lines? in letsencrypt-auto?

**Comment by kuba** [comment]
What do you mean by disabling it? Why do you want to do that?

**Comment by fire** [comment]
> Augeas is only necessary for apache plugin - maybe we could simply not install apache plugin in case of your platform?

Is there a non invasive way of doing this?

**Comment by kuba** [comment]
Oh, I'm really sorry; lost context a bit! Does running `letsencrypt-auto` fail if Augeas (and other apache deps) are not installed?

**Comment by fire** [comment]
> Oh, I'm really sorry; lost context a bit! Does running letsencrypt-auto fail if Augeas (and other apache deps) are not installed?

Yes.

**Comment by Edgewise** [comment]
I got this working on my smartos server running an nginx reverse proxy. I have python2.7 installed and the script recognized it correctly (I just started/pulled today). You might try pulling the latest source and trying again.

I used the following steps:
* commented out the two lines in the letsencrypt-auto file that contain 'letsencrypt-apache'. 
* installed virtualenv and pip using 'pkgin in py27-virtualenv-1.11.6 py27-pip-1.5.6'.
* used the standalone option from the beta invite email.
* updated my nginx conf file with the new certificates

I'm not sure if it was needed, but at one point in trying to debug issues, I deleted the .local and .cache files from my home directory to start fresh. 

Once it ran, the UI was a mess of blue and white striped lines (offset, not running edge to edge) with some text in places. I'm probably missing something like ncurses. I hit enter a few times and it seemed to get me through it. I looked at the log and I guess it was looking for an email address at one point. I believe I had to rerun it to get the Congratulations! message, but the certificates showed up in /etc/letsencrypt/live/... as promised.

My sites are up and running with verified certificates. My hat is off to the developers. Absolutely brilliant! I am not an IT guy or a python developer, so I was very afraid I might be wasting my time.

**Comment by DvdGiessen** [comment]
Just came across the same issue. Augeas is installed on the system, but it seems `certbot` cannot properly find it:
```
root@localhost ~ # certbot-auto certonly --standalone
Checking for new version...
Creating virtual environment...
Installing Python packages...
Installation succeeded.
Requesting root privileges to run certbot...
   /root/.local/share/letsencrypt/bin/letsencrypt certonly --standalone
An unexpected error occurred:
ImportError: Unable to import libaugeas!
Please see the logfile 'certbot.log' for more details.

root@localhost ~ # cat certbot.log
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/bin/letsencrypt", line 11, in <module>
    sys.exit(main())
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/main.py", line 646, in main
    plugins = plugins_disco.PluginsRegistry.find_all()
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/plugins/disco.py", line 174, in find_all
    plugin_ep = PluginEntryPoint(entry_point)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot/plugins/disco.py", line 33, in __init__
    self.plugin_cls = entry_point.load()
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2202, in load
    return self.resolve()
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2208, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot_apache/configurator.py", line 22, in <module>
    from certbot_apache import augeas_configurator
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/certbot_apache/augeas_configurator.py", line 4, in <module>
    import augeas
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/augeas.py", line 78, in <module>
    class Augeas(object):
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/augeas.py", line 82, in Augeas
    _libaugeas = _dlopen("augeas")
  File "/root/.local/share/letsencrypt/lib/python2.7/site-packages/augeas.py", line 75, in _dlopen
    raise ImportError("Unable to import lib%s!" % args[0])
ImportError: Unable to import libaugeas!

root@localhost ~ # ls /opt/local/lib/libaugeas.*
libaugeas.a  libaugeas.la  libaugeas.so  libaugeas.so.0  libaugeas.so.0.19.0
```

Ideally, it should not even try to load the Apache plugin when running standalone.

Another alternative is to just catch and report the error to the user and mark the plugin unavailable (I suppose in `find_all()`?). That would probably be a good solution regardless of platform; `certbot` would not crash on a missing requirement if it wasn't going to be used, and if the user is indeed trying to use the Apache plugin it would nicely report that it failed to load and subsequently cannot be used (until that problem is resolved).

I haven't got much experience in Python, but perhaps someone more knowledgable could work this out. :)

The workaround of removing the certbot-apache and letsencrypt-apache lines from the certbot-auto installer works as well, but that isn't exactly a lasting solution since those changes will probably be overwritten upon any update of the client.

**Comment by bmw** [comment]
@DvdGiessen, we can certainly provide a fix for this problem. I'll move this issue to the next release

**Comment by bmw** [comment]
Unfortunately, this didn't make this release. I'll assign it to the next one.

**Comment by bmw** [comment]
This was fixed in #3128 and will be included in the next release.
